// react-smoothie-modal/dist/index.js
'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);
var PropTypes = _interopDefault(require('prop-types'));

function createCommonjsModule(fn, module) {
	return module = { exports: {} }, fn(module, module.exports), module.exports;
}

var classnames = createCommonjsModule(function (module) {
/*!
  Copyright (c) 2017 Jed Watson.
  Licensed under the MIT License (MIT), see
  http://jedwatson.github.io/classnames
*/
/* global define */

(function () {

	var hasOwn = {}.hasOwnProperty;

	function classNames () {
		var classes = [];

		for (var i = 0; i < arguments.length; i++) {
			var arg = arguments[i];
			if (!arg) continue;

			var argType = typeof arg;

			if (argType === 'string' || argType === 'number') {
				classes.push(arg);
			} else if (Array.isArray(arg) && arg.length) {
				var inner = classNames.apply(null, arg);
				if (inner) {
					classes.push(inner);
				}
			} else if (argType === 'object') {
				for (var key in arg) {
					if (hasOwn.call(arg, key) && arg[key]) {
						classes.push(key);
					}
				}
			}
		}

		return classes.join(' ');
	}

	if (module.exports) {
		classNames.default = classNames;
		module.exports = classNames;
	} else {
		window.classNames = classNames;
	}
}());
});

function styleInject(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') { return; }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css = "/* add css styles here (optional) */\n\n.styles_test__32Qsm {\n  display: inline-block;\n  margin: 2em auto;\n  border: 2px solid #000;\n  font-size: 2em;\n}\n\n.styles_global__1uOe8 {\n  position: fixed;\n  top: 0;\n  left: 0;\n  right: 0;\n  bottom: 0;\n  z-index: 899900;\n  display: block;\n}\n\n.styles_global_inner__3CWqq {\n  position: fixed;\n  top: calc(15% - 33px);\n  left: calc(15% - 33px);\n  right: calc(15% - 33px);\n  bottom: calc(15% - 33px);\n  background: white;\n  z-index: 899902;\n  font-size: calc(0.8vw + 15px);\n  box-shadow: 0px 0px 100px 20px rgba(0, 0, 0, 0.3);\n  border-radius: 8px;\n}\n\n.styles_global_inner_small__OX7jk {\n  position: fixed;\n  top: calc(15% - 33px);\n  left: calc(30% - 33px);\n  right: calc(30% - 33px);\n  bottom: calc(15% - 33px);\n  background: white;\n  z-index: 899902;\n  font-size: calc(0.8vw + 15px);\n  box-shadow: 0px 0px 100px 20px rgba(0, 0, 0, 0.3);\n  border-radius: 8px;\n}\n\n.styles_global_background__AfYA0 {\n  position: fixed;\n  top: 0;\n  left: 0;\n  right: 0;\n  bottom: 0;\n  background: rgba(255, 255, 255, 0.88);\n  z-index: 899901;\n}\n\n.styles_global_close_button__1dNnM {\n  position: absolute;\n  right: 0;\n  background: transparent;\n  cursor: pointer;\n  margin: 12px;\n  width: 32px;\n  height: 32px;\n}\n\n.styles_global_close_image__QPSUr {\n  width: 18px;\n}\n\n.styles_overlay_animate_fade__1J8CI {\n  -webkit-transform-origin: bottom center;\n  transform-origin: bottom center;\n  -webkit-animation: styles_fade-in-pulse-08__1O3S3 0.3s forwards\n    cubic-bezier(0.8, 0.02, 0.45, 0.91);\n  animation: styles_fade-in-pulse-08__1O3S3 0.3s forwards cubic-bezier(0.8, 0.02, 0.45, 0.91);\n}\n\n@keyframes styles_fade-in-pulse-08__1O3S3 {\n  0% {\n    opacity: 0;\n    -webkit-transform: scale(0.8);\n    transform: scale(0.8);\n  }\n\n  50% {\n    opacity: 1;\n  }\n\n  100% {\n    -webkit-transform: scale(1);\n    transform: scale(1);\n  }\n}\n";
var styles = { "test": "styles_test__32Qsm", "global": "styles_global__1uOe8", "global_inner": "styles_global_inner__3CWqq", "global_inner_small": "styles_global_inner_small__OX7jk", "global_background": "styles_global_background__AfYA0", "global_close_button": "styles_global_close_button__1dNnM", "global_close_image": "styles_global_close_image__QPSUr", "overlay_animate_fade": "styles_overlay_animate_fade__1J8CI", "fade-in-pulse-08": "styles_fade-in-pulse-08__1O3S3" };
styleInject(css);

var _ref =
/*#__PURE__*/
React__default.createElement("path", {
  fill: "#4f5458",
  d: "M32.526 2.828L29.698 0 16.263 13.435 2.828 0 0 2.828l13.435 13.435L0 29.698l2.828 2.828 13.435-13.435 13.435 13.435 2.828-2.828-13.435-13.435z"
});

var Icon = "data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20xmlns%3Axlink%3D%22http%3A%2F%2Fwww.w3.org%2F1999%2Fxlink%22%20version%3D%221.1%22%20id%3D%22Capa_1%22%20x%3D%220px%22%20y%3D%220px%22%20viewBox%3D%220%200%2032.526%2032.526%22%20style%3D%22enable-background%3Anew%200%200%2032.526%2032.526%3B%22%20xml%3Aspace%3D%22preserve%22%20width%3D%2229%22%20height%3D%2229%22%3E%3Cpolygon%20points%3D%2232.526%2C2.828%2029.698%2C0%2016.263%2C13.435%202.828%2C0%200%2C2.828%2013.435%2C16.263%200%2C29.698%202.828%2C32.526%2016.263%2C19.091%20%20%2029.698%2C32.526%2032.526%2C29.698%2019.091%2C16.263%20%22%20fill%3D%22%234f5458%22%2F%3E%3C%2Fsvg%3E";

var classCallCheck = function (instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
};

var createClass = function () {
  function defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
      var descriptor = props[i];
      descriptor.enumerable = descriptor.enumerable || false;
      descriptor.configurable = true;
      if ("value" in descriptor) descriptor.writable = true;
      Object.defineProperty(target, descriptor.key, descriptor);
    }
  }

  return function (Constructor, protoProps, staticProps) {
    if (protoProps) defineProperties(Constructor.prototype, protoProps);
    if (staticProps) defineProperties(Constructor, staticProps);
    return Constructor;
  };
}();

var inherits = function (subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });
  if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
};

var possibleConstructorReturn = function (self, call) {
  if (!self) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return call && (typeof call === "object" || typeof call === "function") ? call : self;
};

var SmoothieModal = function (_Component) {
  inherits(SmoothieModal, _Component);

  function SmoothieModal() {
    var _ref;

    var _temp, _this, _ret;

    classCallCheck(this, SmoothieModal);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = possibleConstructorReturn(this, (_ref = SmoothieModal.__proto__ || Object.getPrototypeOf(SmoothieModal)).call.apply(_ref, [this].concat(args))), _this), _this.escFunc = function (event) {
      if (_this.props.escKey) {
        if (event.keyCode === 27) {
          _this.props.handleClose();
        }
      }
    }, _temp), possibleConstructorReturn(_this, _ret);
  }

  createClass(SmoothieModal, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      document.addEventListener('keydown', this.escFunc, false);
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          handleClose = _props.handleClose,
          smallModal = _props.smallModal,
          Backdrop = _props.Backdrop,
          animation = _props.animation,
          extraStyle = _props.extraStyle;


      var Modal = function Modal() {
        var globalAnimation = classnames(styles.global, animation ? styles.overlay_animate_fade : null);
        var extraCss = classnames(smallModal ? styles.global_inner_small : styles.global_inner, extraStyle || null);
        return React__default.createElement(
          'div',
          { className: globalAnimation },
          React__default.createElement('div', { className: styles.global_background, onClick: Backdrop ? handleClose : null }),
          React__default.createElement(
            'div',
            { className: extraCss },
            React__default.createElement(
              'div',
              {
                className: styles.global_close_button,
                onClick: handleClose
              },
              React__default.createElement('img', {
                src: Icon,
                alt: 'Close Modal',
                className: styles.global_close_image
              })
            ),
            children
          )
        );
      };

      return this.props.isShow ? React__default.createElement(Modal, null) : null;
    }
  }]);
  return SmoothieModal;
}(React.Component);

SmoothieModal.propTypes = {
  isShow: PropTypes.bool,
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]).isRequired,
  handleClose: PropTypes.func,
  smallModal: PropTypes.bool,
  Backdrop: PropTypes.bool,
  animation: PropTypes.bool,
  escKey: PropTypes.bool,
  extraStyle: PropTypes.string
};

module.exports = SmoothieModal;
